### PR TITLE
Implementation of TreeEnsemble ai.onnx.ml==5

### PR DESCRIPTION
**Pull request by xadupre**
### Description
Merges PR #21851, #21222.

Implements TreeEnsemble from ai.onnx.ml==5 (CPU).

### CODE REVIEW

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h [hunk 1]
```diff
@@ -957,6 +1089,402 @@ Status TreeEnsembleCommonClassifier<InputType, ThresholdType, OutputType>::compu
...
+    target_class_ids_old.push_back(leaf_targetids[curr_id]);
+    target_class_nodeids_old.push_back(curr_nodeid);
+    target_class_treeids_old.push_back(curr_treeid);
+    target_class_weights_as_tensor_old.push_back(leaf_weights[curr_id]);
+
+    // the below are irrelevant for a `LEAF`
+    nodes_featureids_old.push_back(-1);
+    nodes_truenodeids_old.push_back(-1);
+    nodes_falsenodeids_old.push_back(-1);
+    nodes_values_as_tensor_old.push_back(-1);
```
Comment (bili2002) [inline onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:1280]:
```suggestion
    nodes_featureids_old.push_back(0);
    nodes_truenodeids_old.push_back(0);
    nodes_falsenodeids_old.push_back(0);
    nodes_values_as_tensor_old.push_back(0);
```

Even for leaves, these values can be used as an index in `ProcessTreeNodeLeave`, so they shouldn't be negative.

**Comment by xadupre** [comment]
@bili2002, @cbourjau, I merged the two PR #21851, #21222 to understand the issue with the build. I fixed two, one with the minimal build (a configuration minizing the binary size) and another one because one parameter to AddNodes was not a reference. Feel free to propagate this changes to the existing PRs or to approve this one. I'm ok with both. I'd like to merge them this week.

**Review by bili2002 — APPROVED** [review]
Everything looks great to me here, thank you for the help. Also, I made the changes to the other PR-s. So you can merge this one or the others, whatever is more convenient for you.

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h [hunk 2]
```diff
@@ -957,6 +1089,402 @@ Status TreeEnsembleCommonClassifier<InputType, ThresholdType, OutputType>::compu
   return Status::OK();
 }
 
+template <typename IOType, typename ThresholdType>
+class TreeEnsembleCommonV5 : public TreeEnsembleCommon<IOType, ThresholdType, IOType> {
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:1093]: General comment for this class. It has no data, should the functions be static?
Also, the function signatures are huge, it is hard to read and probably expensive to call. Is it possible, to bind at
least some of those arguments to the instance so they are not passed as args each time?

**Review by yuslepukhin — COMMENTED** [review]
Lots of memory allocations. I had to rewerite a couple kernels in the past due to high allocation contentions that actually affected performance. Is it possible to use InlinedVector at least in some instances?

Using spans in function signagures would enable taking both kinds of vectors as inputs.

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_helper.cc [hunk 3]
```diff
@@ -5,63 +5,53 @@
...
+
+  result = utils::UnpackTensor<T>(proto, std::filesystem::path(), data.data(), tensor_size);
+  ORT_ENFORCE(result.IsOK(), "TreeEnsemble could not unpack tensor attribute ", name);
+
   return Status::OK();
 }
 
 Status GetVectorAttrsOrDefault(const OpKernelInfo& info, const std::string& name, std::vector<double>& data) {
-  return GetVectorAttrsOrDefault(info, name, ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_DOUBLE, data);
+  return _GetVectorAttrsOrDefault(info, name, data);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/tree_ensemble_helper.cc:42]: Please, remove leading underscores in names, this is prohibited.

File: onnxruntime/core/providers/cpu/ml/tree_ensemble.cc [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    p_tree_ensemble_ = std::make_unique<detail::TreeEnsembleCommonV5<T, double>>();
+  } else {
+    p_tree_ensemble_ = std::make_unique<detail::TreeEnsembleCommonV5<T, float>>();
+  }
+  ORT_THROW_IF_ERROR(p_tree_ensemble_->Init(info));
+}
+
+template <typename T>
+Status TreeEnsemble<T>::GetRemovableAttributes(InlinedVector<std::string>& removable_attributes) const {
+  InlinedVector<std::string> names{
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/tree_ensemble.cc:37]: Can this be static constexpr const array of string_view?

**Comment by bili2002** [comment]
I don't have permission to push in this branch, can you give me permission?

**Comment by cbourjau** [comment]
Is there anything we can do to push this forward, @xadupre ?

File: onnxruntime/core/providers/cpu/ml/tree_ensemble.cc (hunk 4 above)
Comment (xadupre) [reply onnxruntime/core/providers/cpu/ml/tree_ensemble.cc:37]: This method is defined with this signature in op.kernel.h. The method modifies the input vector.

**Comment by xadupre** [comment]
> Is there anything we can do to push this forward, @xadupre ?

I did a first refactoring today to avoid very long function signatures. There are a couple of other improvments I still need to make.

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_helper.cc (hunk 3 above)
Comment (xadupre) [reply onnxruntime/core/providers/cpu/ml/tree_ensemble_helper.cc]: done

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h [hunk 5]
```diff
@@ -87,13 +87,48 @@ class TreeEnsembleCommon : public TreeEnsembleCommonAttributes {
   void ComputeAgg(concurrency::ThreadPool* ttp, const Tensor* X, Tensor* Y, Tensor* label, const AGG& agg) const;
 
  private:
+  bool CheckIfSubtreesAreEqual(const size_t left_id, const size_t right_id, const int64_t tree_id, const InlinedVector<NODE_MODE>& cmodes,
+                               const InlinedVector<size_t>& truenode_ids, const InlinedVector<size_t>& falsenode_ids, const std::vector<int64_t>& nodes_featureids,
```
[the remark this answers is not in the document]
Comment (xadupre) [reply onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h]: done

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h [hunk 6]
```diff
@@ -87,13 +87,48 @@ class TreeEnsembleCommon : public TreeEnsembleCommonAttributes {
...
+                               const std::vector<ThresholdType>& nodes_values_as_tensor, const std::vector<float>& node_values,
+                               const std::vector<float>& target_class_weights, const std::vector<ThresholdType>& target_class_weights_as_tensor,
+                               const InlinedVector<TreeNodeElementId>& node_tree_ids, InlinedVector<std::pair<TreeNodeElementId, uint32_t>> indices);
   size_t AddNodes(const size_t i, const InlinedVector<NODE_MODE>& cmodes, const InlinedVector<size_t>& truenode_ids,
                   const InlinedVector<size_t>& falsenode_ids, const std::vector<int64_t>& nodes_featureids,
                   const std::vector<ThresholdType>& nodes_values_as_tensor, const std::vector<float>& node_values,
                   const std::vector<int64_t>& nodes_missing_value_tracks_true, std::vector<size_t>& updated_mapping,
-                  int64_t tree_id, const InlinedVector<TreeNodeElementId>& node_tree_ids);
+                  int64_t tree_id, const InlinedVector<TreeNodeElementId>& node_tree_ids, const std::vector<float>& target_class_weights,
```
[the remark this answers is not in the document]
Comment (xadupre) [reply onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h]: done

**Pull request merged**